"""
–¢–µ—Å—Ç—ã –¥–ª—è gRPC —Å–µ—Ä–≤–µ—Ä–∞ Telegram –±–æ—Ç–∞.

–¢–µ—Å—Ç–∏—Ä—É–µ—Ç —Ñ—É–Ω–∫—Ü–∏–æ–Ω–∞–ª—å–Ω–æ—Å—Ç—å –ø–æ–ª—É—á–µ–Ω–∏—è —Å–æ–æ–±—â–µ–Ω–∏–π —á–µ—Ä–µ–∑ gRPC
–∏ –æ—Ç–ø—Ä–∞–≤–∫—É –∏—Ö –≤ Telegram.
"""
import unittest
import asyncio
from unittest.mock import AsyncMock
import sys
import os

# –î–æ–±–∞–≤–ª—è–µ–º –ø—É—Ç—å –∫ –º–æ–¥—É–ª—è–º —Ç–µ–∫—É—â–µ–π –¥–∏—Ä–µ–∫—Ç–æ—Ä–∏–∏
current_dir = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, current_dir)

try:
    from grpc_server import SendMessageServicer, start_grpc_server
    import hook_pb2
    grpc_available = True
except ImportError:
    # –ï—Å–ª–∏ –∏–º–ø–æ—Ä—Ç –Ω–µ —É–¥–∞–µ—Ç—Å—è, tests –±—É–¥—É—Ç skipped
    SendMessageServicer = None
    start_grpc_server = None
    hook_pb2 = None
    grpc_available = False


class TestSendMessageServicer(unittest.IsolatedAsyncioTestCase):
    """–¢–µ—Å—Ç—ã –¥–ª—è SendMessageServicer."""

    def setUp(self):
        """–ü–æ–¥–≥–æ—Ç–æ–≤–∫–∞ –ø–µ—Ä–µ–¥ –∫–∞–∂–¥—ã–º —Ç–µ—Å—Ç–æ–º."""
        self.webhook_send_mock = AsyncMock()
        if grpc_available:
            self.servicer = SendMessageServicer(
                webhook_send_callback=self.webhook_send_mock
            )

    @unittest.skipUnless(grpc_available, "gRPC modules not available")
    def test_servicer_init_with_callback(self):
        """–¢–µ—Å—Ç –∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏–∏ —Å–µ—Ä–≤–∏—Å–∞ —Å callback."""
        self.assertIsNotNone(self.servicer)
        self.assertEqual(self.servicer.webhook_send_callback, self.webhook_send_mock)

    @unittest.skipUnless(grpc_available, "gRPC modules not available")
    def test_servicer_init_without_callback(self):
        """–¢–µ—Å—Ç –∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏–∏ —Å–µ—Ä–≤–∏—Å–∞ –±–µ–∑ callback."""
        servicer = SendMessageServicer(webhook_send_callback=None)
        self.assertIsNone(servicer.webhook_send_callback)

    @unittest.skipUnless(grpc_available, "gRPC modules not available")
    def test_message_creation(self):
        """–¢–µ—Å—Ç —Å–æ–∑–¥–∞–Ω–∏—è —Å–æ–æ–±—â–µ–Ω–∏—è."""
        message = hook_pb2.Message(
            event="push",
            comment="Test commit",
            chat_id=123456789,
            thread_id=0,
            author="john_doe",
            author_url="https://github.com/john_doe",
            rep_name="test-repo",
            rep_url="https://github.com/test/repo"
        )

        self.assertEqual(message.event, "push")
        self.assertEqual(message.chat_id, 123456789)
        self.assertEqual(message.author, "john_doe")

    @unittest.skipUnless(grpc_available, "gRPC modules not available")
    def test_message_serialization(self):
        """–¢–µ—Å—Ç —Å–µ—Ä–∏–∞–ª–∏–∑–∞—Ü–∏–∏ —Å–æ–æ–±—â–µ–Ω–∏—è."""
        message = hook_pb2.Message(
            event="push",
            comment="Test",
            chat_id=123,
            author="user"
        )

        # –°–µ—Ä–∏–∞–ª–∏–∑—É–µ–º
        serialized = message.SerializeToString()
        self.assertIsInstance(serialized, (bytes, bytearray))
        self.assertGreater(len(serialized), 0)

    @unittest.skipUnless(grpc_available, "gRPC modules not available")
    def test_empty_message(self):
        """–¢–µ—Å—Ç –ø—É—Å—Ç–æ–≥–æ —Å–æ–æ–±—â–µ–Ω–∏—è."""
        message = hook_pb2.Message()

        self.assertEqual(message.event, "")
        self.assertEqual(message.chat_id, 0)
        self.assertEqual(message.thread_id, 0)

    @unittest.skipUnless(grpc_available, "gRPC modules not available")
    def test_message_with_special_characters(self):
        """–¢–µ—Å—Ç —Å–æ–æ–±—â–µ–Ω–∏—è —Å–æ —Å–ø–µ—Ü–∏–∞–ª—å–Ω—ã–º–∏ —Å–∏–º–≤–æ–ª–∞–º–∏."""
        message = hook_pb2.Message(
            event="push",
            comment="Fix bug: √±, √©, ‰∏≠Êñá, üöÄ",
            author="Áî®Êà∑Âêç",
            rep_name="—Ä–µ–ø–æ–∑–∏—Ç–æ—Ä–∏–π"
        )

        self.assertIn("√±", message.comment)
        self.assertIn("‰∏≠Êñá", message.comment)
        self.assertIn("üöÄ", message.comment)
        self.assertIn("Áî®Êà∑Âêç", message.author)


class TestGRPCServerFunctions(unittest.IsolatedAsyncioTestCase):
    """–¢–µ—Å—Ç—ã –¥–ª—è —Ñ—É–Ω–∫—Ü–∏–π gRPC —Å–µ—Ä–≤–µ—Ä–∞."""

    @unittest.skipUnless(grpc_available, "gRPC modules not available")
    async def test_start_grpc_server_creates_task(self):
        """–¢–µ—Å—Ç —á—Ç–æ start_grpc_server —Å–æ–∑–¥–∞–µ—Ç asyncio –∑–∞–¥–∞—á—É."""
        callback = AsyncMock()
        task = start_grpc_server(
            webhook_send_callback=callback,
            port=50052  # –ò—Å–ø–æ–ª—å–∑—É–µ–º –¥—Ä—É–≥–æ–π –ø–æ—Ä—Ç –¥–ª—è —Ç–µ—Å—Ç–æ–≤
        )

        # –ü—Ä–æ–≤–µ—Ä—è–µ–º —á—Ç–æ —Ñ—É–Ω–∫—Ü–∏—è –≤–æ–∑–≤—Ä–∞—â–∞–µ—Ç –∑–∞–¥–∞—á—É –∏–ª–∏ –æ–±—ä–µ–∫—Ç
        self.assertIsNotNone(task)

        # –û—Ç–º–µ–Ω—è–µ–º –∑–∞–¥–∞—á—É –µ—Å–ª–∏ —ç—Ç–æ Task
        if isinstance(task, asyncio.Task):
            task.cancel()
            try:
                await task
            except asyncio.CancelledError:
                pass


class TestMessageStructure(unittest.TestCase):
    """–¢–µ—Å—Ç—ã —Å—Ç—Ä—É–∫—Ç—É—Ä—ã —Å–æ–æ–±—â–µ–Ω–∏—è."""

    @unittest.skipUnless(grpc_available, "gRPC modules not available")
    def test_message_fields(self):
        """–¢–µ—Å—Ç —á—Ç–æ —Å–æ–æ–±—â–µ–Ω–∏–µ –∏–º–µ–µ—Ç –≤—Å–µ –Ω–µ–æ–±—Ö–æ–¥–∏–º—ã–µ –ø–æ–ª—è."""
        fields = [
            'event',
            'comment',
            'chat_id',
            'thread_id',
            'author',
            'author_url',
            'rep_name',
            'rep_url'
        ]

        message = hook_pb2.Message()

        for field_name in fields:
            self.assertTrue(
                hasattr(message, field_name),
                f"Message should have field '{field_name}'"
            )

    @unittest.skipUnless(grpc_available, "gRPC modules not available")
    def test_message_field_types(self):
        """–¢–µ—Å—Ç —Ç–∏–ø–æ–≤ –ø–æ–ª–µ–π —Å–æ–æ–±—â–µ–Ω–∏—è."""
        message = hook_pb2.Message(
            event="push",
            comment="test",
            chat_id=123,
            thread_id=0,
            author="user",
            author_url="url",
            rep_name="repo",
            rep_url="repo_url"
        )

        # –ü—Ä–æ–≤–µ—Ä—è–µ–º —Ç–∏–ø—ã
        self.assertIsInstance(message.event, str)
        self.assertIsInstance(message.comment, str)
        self.assertIsInstance(message.chat_id, int)
        self.assertIsInstance(message.thread_id, int)
        self.assertIsInstance(message.author, str)
        self.assertIsInstance(message.author_url, str)
        self.assertIsInstance(message.rep_name, str)
        self.assertIsInstance(message.rep_url, str)


class TestGRPCIntegration(unittest.TestCase):
    """–ò–Ω—Ç–µ–≥—Ä–∞—Ü–∏–æ–Ω–Ω—ã–µ —Ç–µ—Å—Ç—ã –¥–ª—è gRPC."""

    @unittest.skipUnless(grpc_available, "gRPC modules not available")
    def test_grpc_servicer_with_message(self):
        """–¢–µ—Å—Ç —Å–µ—Ä–≤–∏—Å–∞ —Å —Ä–∞–∑–ª–∏—á–Ω—ã–º–∏ —Å–æ–æ–±—â–µ–Ω–∏—è–º–∏."""
        callback = AsyncMock()
        servicer = SendMessageServicer(webhook_send_callback=callback)

        messages = [
            hook_pb2.Message(event="push", author="user1", chat_id=111),
            hook_pb2.Message(event="pull_request", author="user2", chat_id=222),
            hook_pb2.Message(event="issues", author="user3", chat_id=333),
        ]

        for msg in messages:
            self.assertIsNotNone(msg)
            self.assertGreater(len(msg.event), 0)


class TestGRPCErrorHandling(unittest.TestCase):
    """–¢–µ—Å—Ç—ã –æ–±—Ä–∞–±–æ—Ç–∫–∏ –æ—à–∏–±–æ–∫ –≤ gRPC."""

    @unittest.skipUnless(grpc_available, "gRPC modules not available")
    def test_servicer_without_callback_handles_gracefully(self):
        """–¢–µ—Å—Ç —á—Ç–æ —Å–µ—Ä–≤–∏—Å –æ–±—Ä–∞–±–∞—Ç—ã–≤–∞–µ—Ç –æ—Ç—Å—É—Ç—Å—Ç–≤–∏–µ callback."""
        servicer = SendMessageServicer(webhook_send_callback=None)
        self.assertIsNotNone(servicer)

    @unittest.skipUnless(grpc_available, "gRPC modules not available")
    def test_message_with_zero_values(self):
        """–¢–µ—Å—Ç —Å–æ–æ–±—â–µ–Ω–∏—è —Å –Ω—É–ª–µ–≤—ã–º–∏ –∑–Ω–∞—á–µ–Ω–∏—è–º–∏."""
        message = hook_pb2.Message(
            event="",
            comment="",
            chat_id=0,
            thread_id=0,
            author="",
            author_url="",
            rep_name="",
            rep_url=""
        )

        self.assertEqual(message.event, "")
        self.assertEqual(message.chat_id, 0)

    @unittest.skipUnless(grpc_available, "gRPC modules not available")
    def test_message_with_large_values(self):
        """–¢–µ—Å—Ç —Å–æ–æ–±—â–µ–Ω–∏—è —Å –±–æ–ª—å—à–∏–º–∏ –∑–Ω–∞—á–µ–Ω–∏—è–º–∏."""
        large_string = "x" * 10000
        large_number = 99999999999

        message = hook_pb2.Message(
            event="push",
            comment=large_string,
            chat_id=large_number,
            author=large_string
        )

        self.assertEqual(len(message.comment), 10000)
        self.assertEqual(message.chat_id, large_number)


if __name__ == "__main__":
    unittest.main()
